ci: optimize caching further

thomaseizinger opened the pull request:
This patch-set aims to make several improvements to our CI caching:

1. Use of registry as build cache: Pushes a separate image to our docker registry at GCP that contains the cache layers. This happens for every PR & main. As a result, we can restore from **both** which should make repeated runs of CI on an individual PR faster and give us a good baseline cache for new PRs from `main`. See https://docs.docker.com/build/ci/github-actions/cache/#registry-cache for details. As a nice side-effect, this allows us to use the 10 GB we have on GitHub actions for other jobs.
2. We make better use of `restore-keys` by also attempting to restore the cache if the fingerprint of our lockfiles doesn't match. This is useful for CI runs that upgrade dependencies. Those will restore a cache that is still useful although doesn't quite match. That is better[^1] than not hitting the cache at all.
3. There were two tiny bugs in our Swift and Android builds:
a. We used `rustup show` in the wrong directory and thus did not actually install the toolchain properly.
b. We used `shared-key` instead of `key` for the https://github.com/Swatinem/rust-cache action and thus did not differentiate between jobs properly.
5. Our Dockerfile for Rust had a bug where it did not copy in the `rust-toolchain.toml` file in the `chef` layer and thus also did not use the correctly toolchain.
6. We remove the dedicated gradle cache because the build action already comes with a cache configuration: https://github.com/firezone/firezone/actions/runs/6416847209/job/17421412150#step:10:25

[^1]: Over time, this may mean that our caches grow a bit. In an ideal world, we automatically remove files from the caches that haven't been used in a while. The cache action we use for Rust does that automatically: https://github.com/Swatinem/rust-cache?tab=readme-ov-file#cache-details. As a workaround, we can just purge all caches every now and then.

thomaseizinger committed:
Use registry cache backend

thomaseizinger committed:
Merge branch 'main' into use-registry-as-cache

thomaseizinger committed:
Use max mode

thomaseizinger commented on .github/workflows/ci.yml:
```diff
@@ -101,12 +101,12 @@ jobs:
           cache-to:
             ${{ github.ref == 'refs/heads/main' &&
-            format('type=gha,mode=max,scope={0}', matrix.cache_scope) || '' }}
+            type=registry,ref=${{ env.GCP_REGISTRY }}/${{ env.GCP_PROJECT }}/firezone/${{ matrix.image_name }}:buildcache,mode=max }}
```
Actually, this is probably bad syntax!

jamilbk commented on .github/workflows/ci.yml:
```diff
@@ -101,12 +101,12 @@ jobs:
           platforms: linux/amd64
           build-args: ${{ matrix.build-args }}
           context: ${{ matrix.cache_scope }}/
```
I think we can clean up this var name now too:
```suggestion
          context: ${{ matrix.context }}/
```

thomaseizinger replied:
Can't see `matrix.context` being defined for any of the entries? Do you mean `image_name` perhaps?

jamilbk replied:
`cache_scope` -> `context` in the matrix definition here:
https://github.com/firezone/firezone/pull/2246/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fL37

jamilbk replied:
`s/cache_scope/context/` probably suffices 🎯

thomaseizinger committed:
Use proper syntax

thomaseizinger committed:
Remove stray space

thomaseizinger committed:
Rename `cache_scope` to `context`

thomaseizinger committed:
Always push cache of ref, use cache from main and current ref

thomaseizinger committed:
Merge branch 'main' into use-registry-as-cache

thomaseizinger commented on .github/workflows/ci.yml:
```diff
@@ -76,14 +76,14 @@ jobs:
+          context: ${{ matrix.context }}/
+          cache-from: |
+            type=registry,ref=${{ env.GCP_REGISTRY }}/${{ env.GCP_PROJECT }}/firezone/${{ matrix.image_name }}:buildcache:${{ github.ref }}
+            type=registry,ref=${{ env.GCP_REGISTRY }}/${{ env.GCP_PROJECT }}/firezone/${{ matrix.image_name }}:buildcache:refs/heads/main
```
This is quite neat. We can load the cache from current ref AND main.

thomaseizinger committed:
Don't use double colon

thomaseizinger committed:
Use ref as tag

thomaseizinger committed:
Use ref_name

thomaseizinger committed:
Sanitize ref_name

thomaseizinger committed:
Remove one `/firezone`

thomaseizinger committed:
Fix references

thomaseizinger committed:
Bring back the cache-tag

thomaseizinger committed:
Login to access caches

thomaseizinger committed:
Use CI build caches locally

thomaseizinger committed:
Remove newline in middle of command

thomaseizinger committed:
De-duplicate docker login

thomaseizinger committed:
Merge branch 'main' into use-registry-as-cache

thomaseizinger committed:
Remove unnecessary newline

thomaseizinger committed:
Fix local action ref

thomaseizinger committed:
Fix output value in action

thomaseizinger committed:
Expand permissions

thomaseizinger commented on docker-compose.yml:
```diff
@@ -52,6 +52,8 @@ services:
   web:
     build:
       context: elixir
+      cache_from:
+        - type=registry,ref=us-east1-docker.pkg.dev/firezone-staging/firezone/web-buildcache:main
```
Allows us to transparently reuse caches created by CI :partying_face: 

I tested this by changing the ref to the one from this branch and it worked like a charm! Pretty much instant build locally because it was able to download all the intermediary layers from the registry!

jamilbk replied:
Thanks for implementing this, I think @AndrewDryga had a similar suggestion. Seems like the best approach for now.

thomaseizinger committed:
Checkout code to use local action

thomaseizinger committed:
Also use elixir cache

thomaseizinger committed:
Copy rust-toolchain in chef layer

thomaseizinger commented on rust/Dockerfile:
```diff
@@ -1,4 +1,8 @@
 FROM lukemathwalker/cargo-chef:latest-rust-1.72-slim-bookworm as chef
+
+# See https://github.com/LukeMathWalker/cargo-chef/issues/231.
+COPY rust-toolchain.toml rust-toolchain.toml
```
This is super important for `cargo chef` to work properly.

jamilbk replied:
Ah, I had suspected something may have been off here.

thomaseizinger committed:
Use verbose logging for swift build

thomaseizinger committed:
Run rustup show in correct directory to install toolchain

thomaseizinger committed:
Don't break cache keys half-way, makes them hard to read

thomaseizinger committed:
Remove lock files from restore-keys

Restore-keys allow you to restore for partial cache hits. Changes to the lockfile likely still give you a useful cache.

thomaseizinger committed:
Test swift and kotlin cache restores

thomaseizinger committed:
Use restore keys properly

thomaseizinger committed:
Don't use `shared-key` for swift and android

thomaseizinger committed:
Remove dedicated gradle cache

thomaseizinger committed:
Split cache by apple platform

thomaseizinger committed:
Simplify Rust CI

thomaseizinger committed:
Reference matrix variable that actually exists

thomaseizinger committed:
Only save caches on `main`

jamilbk commented on .github/workflows/rust.yml:
```diff
@@ -67,6 +60,5 @@ jobs:
           workspaces: ./rust
-          shared-key: ${{ runner.os }}
-          save-if: ${{ github.ref == 'refs/heads/main' }}
+          # save-if: ${{ github.ref == 'refs/heads/main' }}
```
```suggestion
          save-if: ${{ github.ref == 'refs/heads/main' }}
```

jamilbk committed:
Uncomment `save-if` to save cache on main

Signed-off-by: Jamil <jamilbk@users.noreply.github.com>
